deep-learning/src/main/scala/com/microsoft/ml/spark/onnx/ONNXModel.scala

```diff
@@ -0,0 +1,496 @@
+        val softmaxCol = df.schema(input).dataType match {
+          case ArrayType(_: NumericType, _) =>
+            val softmaxUdf = UDFUtils.oldUdf({
+              array: Seq[Double] =>
+                val data = BDV(array: _*)
+                (data - softmax(data)).mapValues(math.exp).toSpark
+            }, VectorType)
+            softmaxUdf(col(input).cast(ArrayType(DoubleType)))
+          case MapType(_: NumericType, _: NumericType, _) =>
+            val softmaxUdf = UDFUtils.oldUdf({
+              map: Map[Double, Double] =>
+                val data = BDV(map.toSeq.sortBy(_._1).map(_._2): _*)
+                (data - softmax(data)).mapValues(math.exp).toSpark
+            }, VectorType)
+            softmaxUdf(col(input).cast(MapType(DoubleType, DoubleType)))
+        }
+
+        df.withColumn(output, softmaxCol)
+    }
+  }
+
+  private def argMaxTransform(input: DataFrame): DataFrame = {
+    this.getArgMaxDict.foldLeft(input) {
+      case (df, (input, output)) =>
+        val argmaxCol = df.schema(input).dataType match {
+          case ArrayType(_: NumericType, _) =>
+            val argmaxUdf = UDFUtils.oldUdf({
+              array: Seq[Double] => argmax(array.toArray).toDouble
+            }, DoubleType)
+            argmaxUdf(col(input).cast(ArrayType(DoubleType)))
+          case MapType(_: NumericType, _: NumericType, _) =>
+            val argmaxUdf = UDFUtils.oldUdf({
+              map: Map[Double, Double] =>
+                map.maxBy(_._2)._1
+            }, DoubleType)
+            argmaxUdf(col(input).cast(MapType(DoubleType, DoubleType)))
+        }
+
+        df.withColumn(output, argmaxCol)
+    }
+  }
+
+  private def coerceBatchedDf(df: DataFrame): (DataFrame, Map[String, String]) = {
+    val toArray = UDFUtils.oldUdf({
+      (vectors: Seq[Vector]) => vectors.map(_.toArray)
+    }, ArrayType(ArrayType(DoubleType)))
+
+    this.modelInput.mapValues(f => ArrayType(mapValueInfoToDataType(f.getInfo)))
+      .foldLeft((df, this.getFeedDict)) {
+        case ((accDf, feedDict), (onnxInputName, dataType)) =>
+          val originalColName = this.getFeedDict(onnxInputName)
+          val coercedColName = DatasetExtensions.findUnusedColumnName(originalColName, accDf)
+          val originalCol = df.schema(originalColName).dataType match {
+            case ArrayType(VectorType, _) => toArray(col(originalColName))
+            case _ => col(originalColName)
+          }
+
+          (
+            accDf.withColumn(coercedColName, originalCol.cast(dataType)),
+            feedDict.updated(onnxInputName, coercedColName)
+          )
+      }
+  }
+
+  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)
+
+  override def transformSchema(schema: StructType): StructType = {
+    transformValidateSchema(schema, includePostProcessFields = true)
+  }
+
+  //noinspection ScalaStyle
+  private def transformValidateSchema(schema: StructType, includePostProcessFields: Boolean): StructType = {
```

Comment:
is this ever called without includePostProcessFields?

Reply:
Yes. This is called without includePostProcessFields when we try to obtain an `Encoder` instance (from the output schema) but without the additional post-processing fields. The `Encoder` instance without post-processing field is used for the mapPartition step where we apply the ONNX model, but the post-processing hasn't kicked in yet. I'll change the design a little bit - a private method that returns "ONNX output schema", and the `transformSchema` method will append additional fields to that schema. This will make the design cleaner.

Reply:
Refactored the code to make it cleaner.